quantum/via.c

```diff
@@ -529,92 +622,168 @@ void via_qmk_rgblight_set_value(uint8_t *data) {
-    RGBLIGHT_SPEED_3 = UINT8_MAX / 4 * 3, // UINT8_MAX is really fast
-};
-
-static uint8_t speed_from_rgblight(uint8_t rgblight_speed) {
-    switch (rgblight_speed) {
-        case 0:
-            return RGBLIGHT_SPEED_0;
-        case 1:
-            return RGBLIGHT_SPEED_1;
-        case 2:
-        default:
-            return RGBLIGHT_SPEED_2;
-        case 3:
-            return RGBLIGHT_SPEED_3;
-    }
-}
+void via_qmk_rgb_matrix_command(uint8_t *data, uint8_t length) {
+    // data = [ command_id, channel_id, value_id, value_data ]
+    uint8_t *command_id        = &(data[0]);
+    uint8_t *value_id_and_data = &(data[2]);
 
-static uint8_t speed_to_rgblight(uint8_t rgb_matrix_speed) {
-    if (rgb_matrix_speed < ((RGBLIGHT_SPEED_0 + RGBLIGHT_SPEED_1) / 2)) {
-        return 0;
-    } else if (rgb_matrix_speed < ((RGBLIGHT_SPEED_1 + RGBLIGHT_SPEED_2) / 2)) {
-        return 1;
-    } else if (rgb_matrix_speed < ((RGBLIGHT_SPEED_2 + RGBLIGHT_SPEED_3) / 2)) {
-        return 2;
-    } else {
-        return 3;
+    switch (*command_id) {
+        case id_custom_set_value: {
+            via_qmk_rgb_matrix_set_value(value_id_and_data);
+            break;
+        }
+        case id_custom_get_value: {
+            via_qmk_rgb_matrix_get_value(value_id_and_data);
+            break;
+        }
+        case id_custom_save: {
+            via_qmk_rgb_matrix_save();
+            break;
+        }
+        default: {
+            *command_id = id_unhandled;
+            break;
+        }
     }
 }
 
 void via_qmk_rgb_matrix_get_value(uint8_t *data) {
+    // data = [ value_id, value_data ]
     uint8_t *value_id   = &(data[0]);
     uint8_t *value_data = &(data[1]);
+
     switch (*value_id) {
-        case id_qmk_rgblight_brightness:
+        case id_qmk_rgblight_brightness: {
             value_data[0] = ((uint16_t)rgb_matrix_get_val() * UINT8_MAX) / RGB_MATRIX_MAXIMUM_BRIGHTNESS;
             break;
-        case id_qmk_rgblight_effect:
-            value_data[0] = rgb_matrix_get_mode();
+        }
+        case id_qmk_rgb_matrix_effect: {
+            value_data[0] = rgb_matrix_is_enabled() ? rgb_matrix_get_mode() : 0;
             break;
-        case id_qmk_rgblight_effect_speed:
-            value_data[0] = speed_to_rgblight(rgb_matrix_get_speed());
+        }
+        case id_qmk_rgb_matrix_effect_speed: {
+            value_data[0] = rgb_matrix_get_speed();
             break;
-        case id_qmk_rgblight_color:
+        }
+        case id_qmk_rgb_matrix_color: {
             value_data[0] = rgb_matrix_get_hue();
             value_data[1] = rgb_matrix_get_sat();
             break;
+        }
     }
 }
 
 void via_qmk_rgb_matrix_set_value(uint8_t *data) {
+    // data = [ value_id, value_data ]
     uint8_t *value_id   = &(data[0]);
     uint8_t *value_data = &(data[1]);
     switch (*value_id) {
-        case id_qmk_rgblight_brightness:
+        case id_qmk_rgblight_brightness: {
```

Comment:
Here too.

Reply:
Thanks, probably happened during the merge.